File: Nololiyt.IndexedDbSharp.Raw/Nololiyt.IndexedDbSharp.Raw/CSharp/Async/Implementation/Wrapping/WrappedIdbObjectStore.cs
using Microsoft.Extensions.Options;
using Microsoft.JSInterop;
using Nololiyt.IndexedDbSharp.Raw.CSharp.Async.Wrapping;
using Nololiyt.IndexedDbSharp.Raw.CSharp.Entities;
using System;
using System.Xml.Linq;

namespace Nololiyt.IndexedDbSharp.Raw.CSharp.Async.Implementation.Wrapping
{
    internal sealed class WrappedIdbObjectStore :  WrappedWrappedJsObjectBase, IWrappedIdbObjectStore
    {
        public WrappedIdbObjectStore(IJSObjectReference wrappedObject) : base(wrappedObject)
        {
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKey>> AddAsync<TValue, TResultKey>(TValue value)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("add", value);
            return new WrappedIdbRequestOfValue<TResultKey>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKey>> AddAsync<TValue, TKey, TResultKey>(TValue value, TKey key)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("add", value, key);
            return new WrappedIdbRequestOfValue<TResultKey>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfUndefined> ClearAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("clear");
            return new WrappedIdbRequestOfUndefined(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<long>> CountAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("count");
            return new WrappedIdbRequestOfValue<long>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<long>> CountAsync<TKey>(TKey query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("count", query);
            return new WrappedIdbRequestOfValue<long>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<long>> CountAsync(IdbKeyRangeInfo query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("count", query);
            return new WrappedIdbRequestOfValue<long>(result);
        }

        public async ValueTask<IWrappedIdbIndex> CreateIndexAsync(string name, IdbKeyPath keyPath)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "createIndex", name, keyPath);
            return new WrappedIdbIndex(result);
        }

        public async ValueTask<IWrappedIdbIndex> CreateIndexAsync(
            string name, IdbKeyPath keyPath, IdbIndexParameters options)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "createIndex", name, keyPath, options);
            return new WrappedIdbIndex(result);
        }

        public async ValueTask<IWrappedIdbRequestOfUndefined> DeleteAsync<TKey>(TKey query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "delete", query);
            return new WrappedIdbRequestOfUndefined(result);
        }

        public async ValueTask<IWrappedIdbRequestOfUndefined> DeleteAsync(IdbKeyRangeInfo query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "delete", query);
            return new WrappedIdbRequestOfUndefined(result);
        }

        public async ValueTask DeleteIndexAsync(string name)
        {
            await this.WrappedObject.InvokeVoidAsync("deleteIndex", name);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultElement[]>> GetAllAsync<TResultElement>()
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("getAll");
            return new WrappedIdbRequestOfValue<TResultElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultElement[]>>
            GetAllAsync<TResultElement, TKey>(TKey? query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("getAll", query);
            return new WrappedIdbRequestOfValue<TResultElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultElement[]>> GetAllAsync<TResultElement>(
            IdbKeyRangeInfo? query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("getAll", query);
            return new WrappedIdbRequestOfValue<TResultElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultElement[]>> GetAllAsync<TResultElement>(
            int count)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "getAllSkipQuery", count);
            return new WrappedIdbRequestOfValue<TResultElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultElement[]>>
            GetAllAsync<TResultElement, TKey>(TKey? query, int count)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "getAll", query, count);
            return new WrappedIdbRequestOfValue<TResultElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultElement[]>>
            GetAllAsync<TResultElement>(IdbKeyRangeInfo? query, int count)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "getAll", query, count);
            return new WrappedIdbRequestOfValue<TResultElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKeyElement[]>>
            GetAllKeysAsync<TResultKeyElement>()
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("getAllKeys");
            return new WrappedIdbRequestOfValue<TResultKeyElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKeyElement[]>>
            GetAllKeysAsync<TKey, TResultKeyElement>(TKey? query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("getAllKeys", query);
            return new WrappedIdbRequestOfValue<TResultKeyElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKeyElement[]>>
            GetAllKeysAsync<TResultKeyElement>(IdbKeyRangeInfo? query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("getAllKeys", query);
            return new WrappedIdbRequestOfValue<TResultKeyElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKeyElement[]>>
            GetAllKeysAsync<TResultKeyElement>(int count)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "getAllKeysSkipQuery", count);
            return new WrappedIdbRequestOfValue<TResultKeyElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKeyElement[]>> 
            GetAllKeysAsync<TKey, TResultKeyElement>(TKey? query, int count)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "getAllKeys", query, count);
            return new WrappedIdbRequestOfValue<TResultKeyElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKeyElement[]>>
            GetAllKeysAsync<TResultKeyElement>(IdbKeyRangeInfo? query, int count)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "getAllKeys", query, count);
            return new WrappedIdbRequestOfValue<TResultKeyElement[]>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TValue>> GetAsync<TValue, TKey>(TKey query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("get", query);
            return new WrappedIdbRequestOfValue<TValue>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TValue>> GetAsync<TValue>(IdbKeyRangeInfo query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("get", query);
            return new WrappedIdbRequestOfValue<TValue>(result);
        }

        public async ValueTask<bool> GetAutoIncrementAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<bool>("autoIncrement");
            return result;
        }

        public async ValueTask<string[]> GetIndexNamesAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<string[]>("indexNames");
            return result;
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKey?>> GetKeyAsync<TKey, TResultKey>(
            TKey query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("getKey", query);
            return new WrappedIdbRequestOfValue<TResultKey?>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKey?>> GetKeyAsync<TResultKey>(IdbKeyRangeInfo query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("getKey", query);
            return new WrappedIdbRequestOfValue<TResultKey?>(result);
        }

        public async ValueTask<IdbKeyPath> GetKeyPathAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<IdbKeyPath>("keyPath");
            return result;
        }

        public async ValueTask<IWrappedIdbTransaction> GetTransactionAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("transaction");
            return new WrappedIdbTransaction(result);
        }

        public async ValueTask<IWrappedIdbIndex> IndexAsync(string name)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("index", name);
            return new WrappedIdbIndex(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorWithValueOrNull> OpenCursorAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("openCursor");
            return new WrappedIdbRequestOfIdbCursorWithValueOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorWithValueOrNull>
            OpenCursorAsync<TKey>(TKey? query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("openCursor", query);
            return new WrappedIdbRequestOfIdbCursorWithValueOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorWithValueOrNull>
            OpenCursorAsync(IdbKeyRangeInfo? query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("openCursor", query);
            return new WrappedIdbRequestOfIdbCursorWithValueOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorWithValueOrNull>
            OpenCursorAsync(IdbCursorDirection direction)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "openCursorSkipQuery", direction);
            return new WrappedIdbRequestOfIdbCursorWithValueOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorWithValueOrNull>
            OpenCursorAsync<TKey>(TKey? query, IdbCursorDirection direction)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "openCursor", query, direction);
            return new WrappedIdbRequestOfIdbCursorWithValueOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorWithValueOrNull> OpenCursorAsync(
            IdbKeyRangeInfo? query, IdbCursorDirection direction)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "openCursor", query, direction);
            return new WrappedIdbRequestOfIdbCursorWithValueOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorOrNull> OpenKeyCursorAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "openKeyCursor");
            return new WrappedIdbRequestOfIdbCursorOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorOrNull> OpenKeyCursorAsync<TKey>(TKey? query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "openKeyCursor", query);
            return new WrappedIdbRequestOfIdbCursorOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorOrNull> OpenKeyCursorAsync(IdbKeyRangeInfo? query)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "openKeyCursor", query);
            return new WrappedIdbRequestOfIdbCursorOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorOrNull> OpenKeyCursorAsync(IdbCursorDirection direction)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "openKeyCursorSkipQuery", direction);
            return new WrappedIdbRequestOfIdbCursorOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorOrNull> OpenKeyCursorAsync<TKey>(TKey? query, IdbCursorDirection direction)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "openKeyCursor", query, direction);
            return new WrappedIdbRequestOfIdbCursorOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfIdbCursorOrNull> OpenKeyCursorAsync(IdbKeyRangeInfo? query, IdbCursorDirection direction)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "openKeyCursor", query, direction);
            return new WrappedIdbRequestOfIdbCursorOrNull(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKey>>
            PutAsync<TValue, TResultKey>(TValue value)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "put", value);
            return new WrappedIdbRequestOfValue<TResultKey>(result);
        }

        public async ValueTask<IWrappedIdbRequestOfValue<TResultKey>>
            PutAsync<TValue, TKey, TResultKey>(TValue value, TKey key)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>(
                "put", value, key);
            return new WrappedIdbRequestOfValue<TResultKey>(result);
        }
    }
}

File: Nololiyt.IndexedDbSharp.Raw/Nololiyt.IndexedDbSharp.Raw/Typescripts/Async/Wrapping/WrappedIdbObjectStore.ts
import { convertToIdbKeyRange, IdbKeyRangeInfo, isIdbKeyRangeInfo } from "../Entities/IdbKeyRangeInfo.js";
import { WrappedIdbIndex } from "./WrappedIdbIndex.js";
import { WrappedIdbRequest, WrappedIdbRequestOfAny, WrappedIdbRequestOfAnyArray, WrappedIdbRequestOfIdbCursorOrNull, WrappedIdbRequestOfIdbCursorWithValueOrNull, WrappedIdbRequestOfIdbValidKey, WrappedIdbRequestOfIdbValidKeyArray, WrappedIdbRequestOfIdbValidKeyOrUndefined, WrappedIdbRequestOfNumber, WrappedIdbRequestOfUndefined } from "./WrappedIdbRequests.js";
import { WrappedIdbTransaction } from "./WrappedIdbTransaction.js";

export class WrappedIdbObjectStore
{
    private wrapped;
    constructor(wrapped: IDBObjectStore)
    {
        this.wrapped = wrapped;
    }

    wrappedObject()
    {
        return this.wrapped;
    }

    add(value: any, key?: IDBValidKey | undefined): WrappedIdbRequestOfIdbValidKey
    {
        const result = this.wrapped.add(value, key);
        return new WrappedIdbRequestOfIdbValidKey(result);
    }

    autoIncrement(): boolean
    {
        return this.wrapped.autoIncrement;
    }

    clear(): WrappedIdbRequestOfUndefined
    {
        const result = this.wrapped.clear();
        return new WrappedIdbRequestOfUndefined(result);
    }

    count(query?: IDBValidKey | IdbKeyRangeInfo | undefined): WrappedIdbRequestOfNumber
    {
        let result: IDBRequest<number>;
        if (isIdbKeyRangeInfo(query))
            result = this.wrapped.count(convertToIdbKeyRange(query));
        else
            result = this.wrapped.count(query);
        return new WrappedIdbRequestOfNumber(result);
    }

    createIndex(
        name: string,
        keyPath: string | string[],
        options?: IDBIndexParameters | undefined): WrappedIdbIndex
    {
        const result = this.wrapped.createIndex(name, keyPath, options);
        return new WrappedIdbIndex(result);
    }

    delete(query: IDBValidKey | IdbKeyRangeInfo): WrappedIdbRequestOfUndefined
    {
        let result: IDBRequest<undefined>;
        if (isIdbKeyRangeInfo(query))
            result = this.wrapped.delete(convertToIdbKeyRange(query));
        else
            result = this.wrapped.delete(query);
        return new WrappedIdbRequestOfUndefined(result);
    }

    deleteIndex(name: string): void
    {
        this.wrapped.deleteIndex(name);
    }

    get(query: IDBValidKey | IdbKeyRangeInfo): WrappedIdbRequestOfAny
    {
        let result: IDBRequest<any>;
        if (isIdbKeyRangeInfo(query))
            result = this.wrapped.get(convertToIdbKeyRange(query));
        else
            result = this.wrapped.get(query);
        return new WrappedIdbRequestOfAny(result);
    }

    getAll(
        query?: IDBValidKey | IdbKeyRangeInfo | null | undefined,
        count?: number | undefined): WrappedIdbRequestOfAnyArray
    {
        let result: IDBRequest<any[]>;
        if (isIdbKeyRangeInfo(query))
            result = this.wrapped.getAll(convertToIdbKeyRange(query), count);
        else
            result = this.wrapped.getAll(query, count);
        return new WrappedIdbRequestOfAnyArray(result);
    }
    getAllSkipQuery(count?: number | undefined): WrappedIdbRequestOfAnyArray
    {
        let result: IDBRequest<any[]>;
        result = this.wrapped.getAll(undefined, count);
        return new WrappedIdbRequestOfAnyArray(result);
    }

    getAllKeys(
        query?: IDBValidKey | IdbKeyRangeInfo | null | undefined,
        count?: number | undefined): WrappedIdbRequestOfIdbValidKeyArray
    {
        let result: IDBRequest<IDBValidKey[]>;
        if (isIdbKeyRangeInfo(query))
            result = this.wrapped.getAllKeys(convertToIdbKeyRange(query), count);
        else
            result = this.wrapped.getAllKeys(query, count);
        return new WrappedIdbRequestOfIdbValidKeyArray(result);
    }
    getAllKeysSkipQuery(count?: number | undefined): WrappedIdbRequestOfIdbValidKeyArray
    {
        let result: IDBRequest<IDBValidKey[]>;
        result = this.wrapped.getAllKeys(undefined, count);
        return new WrappedIdbRequestOfIdbValidKeyArray(result);
    }

    getKey(query: IDBValidKey | IdbKeyRangeInfo): WrappedIdbRequestOfIdbValidKeyOrUndefined
    {
        let result: IDBRequest<IDBValidKey | undefined>;
        if (isIdbKeyRangeInfo(query))
            result = this.wrapped.getKey(convertToIdbKeyRange(query));
        else
            result = this.wrapped.getKey(query);
        return new WrappedIdbRequestOfIdbValidKeyOrUndefined(result);
    }

    index(name: string): WrappedIdbIndex
    {
        const result = this.wrapped.index(name);
        return new WrappedIdbIndex(result);
    }

    indexNames(): string[]
    {
        const names = this.wrapped.indexNames;
        const result: string[] = [];
        for (const name in names)
            result.push(name);
        return result;
    }

    keyPath(): string | string[]
    {
        return this.wrapped.keyPath;
    }

    openCursor(
        query?: IDBValidKey | IdbKeyRangeInfo | null | undefined,
        direction?: IDBCursorDirection | undefined): WrappedIdbRequestOfIdbCursorWithValueOrNull
    {
        let result: IDBRequest<IDBCursorWithValue | null>;
        if (isIdbKeyRangeInfo(query))
            result = this.wrapped.openCursor(convertToIdbKeyRange(query), direction);
        else
            result = this.wrapped.openCursor(query, direction);
        return new WrappedIdbRequestOfIdbCursorWithValueOrNull(result);
    }
    openCursorSkipQuery(direction?: IDBCursorDirection | undefined): WrappedIdbRequestOfIdbCursorWithValueOrNull
    {
        let result: IDBRequest<IDBCursorWithValue | null>;
        result = this.wrapped.openCursor(undefined, direction);
        return new WrappedIdbRequestOfIdbCursorWithValueOrNull(result);
    }

    openKeyCursor(
        query?: IDBValidKey | IdbKeyRangeInfo | null | undefined,
        direction?: IDBCursorDirection | undefined): WrappedIdbRequestOfIdbCursorOrNull
    {
        let result: IDBRequest<IDBCursor | null>;
        if (isIdbKeyRangeInfo(query))
            result = this.wrapped.openKeyCursor(convertToIdbKeyRange(query), direction);
        else
            result = this.wrapped.openKeyCursor(query, direction);
        return new WrappedIdbRequestOfIdbCursorOrNull(result);
    }
    openKeyCursorSkipQuery(direction?: IDBCursorDirection | undefined): WrappedIdbRequestOfIdbCursorOrNull
    {
        let result: IDBRequest<IDBCursor | null>;
        result = this.wrapped.openKeyCursor(undefined, direction);
        return new WrappedIdbRequestOfIdbCursorOrNull(result);
    }

    put(value: any, key?: IDBValidKey | undefined): WrappedIdbRequestOfIdbValidKey
    {
        const result = this.wrapped.put(value, key);
        return new WrappedIdbRequestOfIdbValidKey(result);
    }

    transaction(): WrappedIdbTransaction
    {
        const result = this.wrapped.transaction;
        return new WrappedIdbTransaction(result);
    }
}